internal/pipe/krew/krew.go

```diff
@@ -0,0 +1,330 @@
+		Metadata: Metadata{
+			Name: cfg.Name,
+		},
+		Spec: Spec{
+			Homepage:         cfg.Homepage,
+			Version:          "v" + ctx.Version,
+			ShortDescription: shortDesc,
+			Description:      desc,
+			Caveats:          cfg.Caveats,
+		},
+	}
+
+	for _, art := range artifacts {
+		sum, err := art.Checksum("sha256")
+		if err != nil {
+			return result, err
+		}
+
+		if cfg.URLTemplate == "" {
+			url, err := cl.ReleaseURLTemplate(ctx)
+			if err != nil {
+				return result, err
+			}
+			cfg.URLTemplate = url
+		}
+		url, err := tmpl.New(ctx).WithArtifact(art, map[string]string{}).Apply(cfg.URLTemplate)
+		if err != nil {
+			return result, err
+		}
+
+		goarch := []string{art.Goarch}
+		if art.Goarch == "all" {
+			goarch = []string{"amd64", "arm64"}
+		}
+
+		for _, arch := range goarch {
+			bins := art.ExtraOr(artifact.ExtraBinaries, []string{}).([]string)
+			if len(bins) != 1 {
+				return result, fmt.Errorf("krew: only one binary per archive allowed, got %d on %q", len(bins), art.Name)
+			}
+			result.Spec.Platforms = append(result.Spec.Platforms, Platform{
+				Bin:    bins[0],
+				URI:    url,
+				Sha256: sum,
+				Selector: Selector{
+					MatchLabels: MatchLabels{
+						Os:   art.Goos,
+						Arch: arch,
+					},
+				},
+			})
+		}
+	}
+
+	return result, nil
+}
+
+// Publish krew manifest.
+func (Pipe) Publish(ctx *context.Context) error {
+	cli, err := client.New(ctx)
+	if err != nil {
+		return err
+	}
+	return publishAll(ctx, cli)
+}
+
+func publishAll(ctx *context.Context, cli client.Client) error {
+	skips := pipe.SkipMemento{}
+	for _, manifest := range ctx.Artifacts.Filter(artifact.ByType(artifact.KrewPluginManifest)).List() {
+		err := doPublish(ctx, manifest, cli)
+		if err != nil && pipe.IsSkip(err) {
+			skips.Remember(err)
+			continue
+		}
+		if err != nil {
+			return err
+		}
+	}
+	return skips.Evaluate()
+}
+
+func doPublish(ctx *context.Context, manifest *artifact.Artifact, cl client.Client) error {
+	cfg := manifest.Extra[krewConfigExtra].(config.Krew)
+	var err error
+	cl, err = client.NewIfToken(ctx, cl, cfg.Index.Token)
+	if err != nil {
+		return err
+	}
+
+	if strings.TrimSpace(cfg.SkipUpload) == "true" {
+		return pipe.Skip("krews.skip_upload is set")
+	}
+
+	if strings.TrimSpace(cfg.SkipUpload) == "auto" && ctx.Semver.Prerelease != "" {
+		return pipe.Skip("prerelease detected with 'auto' upload, skipping krew publish")
+	}
+
+	repo := client.RepoFromRef(cfg.Index)
+
+	gpath := buildManifestPath(manifestsFolder, manifest.Name)
+	log.WithField("manifest", gpath).
+		WithField("repo", repo.String()).
+		Info("pushing")
+
+	msg, err := tmpl.New(ctx).Apply(cfg.CommitMessageTemplate)
+	if err != nil {
+		return err
+	}
+
+	content, err := os.ReadFile(manifest.Path)
+	if err != nil {
+		return err
+	}
+
+	return cl.CreateFile(ctx, cfg.CommitAuthor, repo, content, gpath, msg)
+}
+
+func buildManifestPath(folder, filename string) string {
+	return path.Join(folder, filename)
+}
+
+type Manifest struct {
```

Comment:
FWIW we offer this as a type in https://pkg.go.dev/sigs.k8s.io/krew/pkg/index#Plugin

Reply:
I saw, didn't use it because it comes with quite a few dependencies we don't need... seemed easy enough to copy the relevant parts